File: broken-calculator.cpp
//Time - O(logY)
//Space - O(1)
class Solution {
public:
    int brokenCalc(int X, int Y) {
        int cnt = 0;
        
        while(Y>X){
            if(Y%2 == 0){
                Y = Y/2;
            }else{
                Y = Y+1;
            }
            cnt++;
        }
        
        cnt = cnt + X-Y;
        return cnt;
        
    }
};

File: assignments/src/Broken_Calculator.java
// Approach: Think backwards - as it gives the correct Greedy approach required to
// solve this problem i.e. divide by 2 if target even or +1 if odd then divide.
// Time: O(log target)
// Space: O(1)

class Broken_Calculator {
    public int brokenCalc(int startValue, int target) {
        int steps = 0;

        while (target > startValue) {
            steps++;

            if (target % 2 == 1) {      // odd
                target++;
            } else {                    // even
                target /= 2;
            }
        }
        return steps + startValue - target;
    }
}